cmd/geoipupdate/main.go

```diff
@@ -52,27 +58,60 @@ func main() {
 
 	client := geoipupdate.NewClient(config)
 
-	if err = run(client, config); err != nil {
+	downloadFunc := func(editionID string) error {
+		return download(client, config, editionID)
+	}
+
+	if err = run(config, downloadFunc); err != nil {
 		fatalLogger("error retrieving updates", err)
 	}
 }
 
-func run(client *http.Client, config *geoipupdate.Config) error {
-	dbReader := database.NewHTTPDatabaseReader(client, config)
-
+// run concurrently downloads GeoIP databases using the provided config.
+// `config.Parallelism` limits the number of concurrent downloads that can be
+// executed and setting it to 1 would just mean databases would download
+// sequencially.
+func run(
```

Comment:
It is unfortunate that we need to do the parallelism here rather than in `pkg/geoipupdate`, but anything else would require a fairly significant refactoring. Maybe we should add a comment there to make it clearer.

Reply:
Yeah, I can see the argument for keeping the concurrency separated generally, but it is being defined in the config for the package and the locking happens there as well.